app/src/main/java/com/example/mytestproject/ui/searchCity/SearchCityFragment.kt

```diff
@@ -0,0 +1,73 @@
+package com.example.mytestproject.ui.searchCity
+
+import android.content.Context
+import android.content.SharedPreferences
+import android.os.Bundle
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import androidx.fragment.app.Fragment
+import androidx.lifecycle.ViewModelProvider
+import androidx.navigation.fragment.findNavController
+import androidx.recyclerview.widget.LinearLayoutManager
+import com.example.mytestproject.App
+import com.example.mytestproject.R
+import com.example.mytestproject.util.CITY_ID
+import com.example.mytestproject.util.SHARED_PREFS
+import kotlinx.android.synthetic.main.fragment_search_city.*
+import javax.inject.Inject
+
+class SearchCityFragment : Fragment() {
+
+    private lateinit var adapter: SearchCityAdapter
+
+    @Inject
+    lateinit var searchCityFactory: SearchCityFactory
+    private lateinit var searchViewModel: SearchCityViewModel
+
+    override fun onCreateView(
+        inflater: LayoutInflater, container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        return inflater.inflate(R.layout.fragment_search_city, container, false)
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        (activity?.applicationContext as App).weatherDataComponent.inject(this)
+
+        searchViewModel =
+            ViewModelProvider(this, searchCityFactory).get(SearchCityViewModel::class.java)
+
+        adapter = SearchCityAdapter()
+
+        rv_search_city.adapter = adapter
+        rv_search_city.layoutManager =
+            LinearLayoutManager(requireContext(), LinearLayoutManager.VERTICAL, false)
+
+        searchViewModel.searchCity(adapter, search_view)
+
+        saveCityId()
+
+    }
+
+    private fun saveCityId() {
+        adapter.onClickItemListener(object :
```

Comment:
There should be no logic in Fragment. Adapter should pass the value to ViewModel, like a viewModel.onCityChosen(id). And ViewModel implements all this pref saving logic. Also ViewModel should ask Fragment to  navigate back.

Reply:
Fixed